core/src/main/scala/org/bykn/bosatsu/Program.scala

```diff
@@ -52,22 +55,50 @@ object Program {
       d.toExpr(nameToType, nameToCons)
 
     def defToT(
-      types: rankn.TypeEnv,
-      d: TypeDefinitionStatement): rankn.TypeEnv =
+      types: TypeEnv,
+      d: TypeDefinitionStatement): TypeEnv =
       types.addDefinedType(d.toDefinition(pn0, nameToType))
 
+    val allNames = stmt.toStream.flatMap {
+      case Bind(BindingStatement(bound, _, _)) => bound.names
+      case _ => Nil
+    }.toSet
+
+    // Each time we need a name, we can call anonNames.next()
+    // it is mutable, but in a limited scope
+    val anonNames = rankn.Type.allBinders.iterator.map(_.name).filterNot(allNames)
 
-    def bindings(b: Pattern[Option[String], TypeRef], decl: Expr[Declaration]): List[(String, Expr[Declaration])] =
+    def bindings(b: Pattern[(PackageName, ConstructorName), Type], decl: Expr[Declaration]): List[(String, Expr[Declaration])] =
       b match {
         case Pattern.Var(nm) => (nm, decl) :: Nil
-        case unsupported => sys.error(s"unsupported pattern: $unsupported")
+        case Pattern.Annotation(p, tpe) =>
+          // we can just move the annotation to the expr:
+          bindings(p, Expr.Annotation(decl, tpe, decl.tag))
+        case complex =>
+          val (prefix, left) = decl match {
+            case v@Expr.Var(_, _, _) =>
+              // no need to make a new var to point to a var
+              (Nil, v)
+            case _ =>
+              val thisName = anonNames.next()
+              val v = Expr.Var(None, thisName, decl.tag)
+              ((thisName, decl) :: Nil, v)
+          }
+
+          val tail = complex.names.map { nm =>
+            val pat = complex.filterVars(Set(nm))
+            (nm, Expr.Match(left,
+              NonEmptyList.of((pat, Expr.Var(None, nm, decl.tag))), decl.tag))
+          }
```

Comment:
If tail is empty we still need to typecheck that the pattern is sane. We probably want to lint no-op pattern matches but we shouldn’t accept a bad match just because the code doesn’t bing any names.